transport/ssh.go

```diff
@@ -0,0 +1,111 @@
+package transport
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/kballard/go-shellquote"
+	"github.com/kevinburke/ssh_config"
+	"github.com/puppetlabs/wash/plugin"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+)
+
+func newAgent() (ssh.AuthMethod, error) {
+	sshAgent, err := net.Dial("unix", os.Getenv("SSH_AUTH_SOCK"))
+
+	if err != nil {
+		return nil, err
+	}
+	return ssh.PublicKeysCallback(agent.NewClient(sshAgent).Signers), nil
+}
+
+func sshConnect(host, port, user string) (*ssh.Client, error) {
+	agent, err := newAgent()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to find config from SSH_AUTH_SOCK: %s", err)
+	}
+
+	sshConfig := &ssh.ClientConfig{
+		User: user,
+		Auth: []ssh.AuthMethod{agent},
+		// TODO: use https://skarlso.github.io/2019/02/17/go-ssh-with-host-key-verification/
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+
+	// TODO: cache connections
+	return ssh.Dial("tcp", host+":"+port, sshConfig)
+}
+
+// Identity identifies how to connect to a target.
+type Identity struct {
+	Host, User string
+}
+
+// ExecSSH executes against a target via SSH. It will look up port, user, and other configuration
+// by exact hostname match from default SSH config files. Identity can be used to override the
+// default user.
+//
+//
+func ExecSSH(ctx context.Context, id Identity, cmd []string, opts plugin.ExecOptions) (plugin.ExecResult, error) {
+	var result plugin.ExecResult
+
+	// find port, username, etc from .ssh/config
+	port, err := ssh_config.GetStrict(id.Host, "Port")
+	if err != nil {
+		return result, err
+	}
+
+	user := id.User
+	if user == "" {
+		if user, err = ssh_config.GetStrict(id.Host, "User"); err != nil {
+			return result, err
+		}
+	}
+
+	if user == "" {
+		user = "root"
+	}
+
+	connection, err := sshConnect(id.Host, port, user)
+	if err != nil {
+		return result, fmt.Errorf("Failed to connect: %s", err)
+	}
+
+	// Run command via session
+	session, err := connection.NewSession()
+	if err != nil {
+		return result, fmt.Errorf("Failed to create session: %s", err)
+	}
+
+	outputch, stdout, stderr := plugin.CreateExecOutputStreams(ctx)
+	session.Stdin, session.Stdout, session.Stderr = opts.Stdin, stdout, stderr
```

Comment:
Needs additional work to support TTY and context completion.

Reply:
Why do we need TTY support? If it's to kill the process, couldn't we just use https://godoc.org/golang.org/x/crypto/ssh#Session.Signal ?

Reply:
Good point. I'd noticed that. I think I could probably always send a signal if still waiting when the context finishes.